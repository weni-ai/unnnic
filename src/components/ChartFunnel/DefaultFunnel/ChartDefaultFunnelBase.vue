<template>
  <section class="unnnic-chart-funnel-base-container">
    <section
      v-for="(step, index) in data"
      :key="index"
      class="unnnic-chart-funnel-base-item"
    >
    <section 
      :class="[
        'overflow-hidden', 
         step.widthClass,
        { 'first-item': index === 0, 'last-item': index === data.length - 1 }
      ]">
      <section
        :class="['unnnic-chart-funnel-base-item__card',  { 'first-item': index === 0}]"
        :style="{ backgroundColor: step.color }"
      ></section>
    </section>

    <section :class="['unnnic-chart-funnel-base-item__text', { 'last-item': index === data.length - 1 }]">
      <section class="unnnic-chart-funnel-base-item__text__values">
        <p class="unnnic-chart-funnel-base-item__text__values-title">
          {{ step.percentage }}
        </p>
        <p class="unnnic-chart-funnel-base-item__text__values-sub-title">
          &nbsp; | {{ step.value }}
        </p>
      </section>
        <p class="unnnic-chart-funnel-base-item__text-description">
          {{ step.description }}
        </p>
      </section>
    </section>
  </section>
</template>

<script setup lang="ts">
interface FunnelStep {
  percentage: number | string;
  value: number | string;
  description: string;
  widthClass: string;
  color: string;
}

defineProps<{
  data: FunnelStep[]
}>();
</script>

<style lang="scss" scoped>
@import '../../../assets/scss/unnnic.scss';

.unnnic-chart-funnel-base-container {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
}

.unnnic-chart-funnel-base-item {
  display: flex;
  align-items: center;
  justify-content: flex-start;
  flex-grow: 1;

  &__card {
    height: 100%;
    transition: background-color 0.3s ease;
    transform: skew(-12deg, 0deg) translateX(-20px);
    border-radius: 0 0 $unnnic-spacing-xs 0;

    &.first-item {
    border-radius: $unnnic-spacing-xs $unnnic-spacing-xs $unnnic-spacing-xs 0;
    }
  }

  &__text {
    display: flex;
    flex-direction: column;
    justify-content: center;
    position: relative;
    padding-left: $unnnic-spacing-sm;
    padding-right: $unnnic-spacing-sm;
    height: 100%;

    &::after {
      content: '';
      position: absolute;
      left: 0;
      bottom: 0px;
      width: 100%;
      height: 1px;
      background-color: $unnnic-color-neutral-soft;
      z-index: -1;
    }

    &.last-item::after {
      display: none;
    }
    
    &-description {
      margin: 0;
      color: $unnnic-color-neutral-dark;
      text-align: center;

      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-md;
      font-style: normal;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;
    }
    &__values {
      display: flex;
      &-title {
        margin: 0;
        color: $unnnic-color-neutral-dark;

        font-family: $unnnic-font-family-secondary;
        font-size: $unnnic-font-size-body-lg;
        font-style: normal;
        font-weight: $unnnic-font-weight-bold;
        line-height: $unnnic-font-size-body-lg + $unnnic-line-height-md;
      }

      &-sub-title {
        display: flex;
        align-items: end;
        margin: 0;
        color: $unnnic-color-neutral-cloudy;

        font-family: $unnnic-font-family-secondary;
        font-size: $unnnic-font-size-body-md;
        font-style: normal;
        font-weight: $unnnic-font-weight-regular;
        line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;
      }
    }
  }
}

.w-60 {
  width: 60%;
}
.w-50 {
  width: 50%;
}
.w-40 {
  width: 40%;
}
.w-30 {
  width: 30%;
}
.w-20 {
  width: 20%;
}
.overflow-hidden {
  height: 100%;
  overflow: hidden;
  position: relative;

  &.first-item {
    border-radius: $unnnic-spacing-xs $unnnic-spacing-xs 0px 0;
    }

  &.last-item {
    border-radius: 0 0 $unnnic-spacing-xs $unnnic-spacing-xs;
  }

  &::after {
    content: '';
    position: absolute;
    left: 0;
    bottom: -1px;
    width: 100%;
    height: 2px;
    background-color: $unnnic-color-neutral-soft;
    z-index: -1;
  }

  &.last-item::after {
    display: none;
  }
}
</style>
