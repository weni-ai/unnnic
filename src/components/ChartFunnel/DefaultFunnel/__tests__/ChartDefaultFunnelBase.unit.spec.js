import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, afterEach, test } from 'vitest';
import ChartDefaultFunnelBase from '@/components/ChartFunnel/DefaultFunnel/ChartDefaultFunnelBase.vue';

describe('ChartDefaultFunnelBase.vue', () => {
  let wrapper;

  const sampleData = [
    {
      percentage: '90%',
      value: 900,
      description: 'Viewed the landing page',
      widthClass: 'w-60',
      color: '#0077B6',
    },
    {
      percentage: '80%',
      value: 800,
      description: 'Clicked on the CTA',
      widthClass: 'w-50',
      color: '#00A8E8',
    },
    {
      percentage: '70%',
      value: 700,
      description: 'Signed up for the trial',
      widthClass: 'w-40',
      color: '#00BFFF',
    },
    {
      percentage: '60%',
      value: 600,
      description: 'Completed the onboarding',
      widthClass: 'w-30',
      color: '#87CEEB',
    },
    {
      percentage: '50%',
      value: 500,
      description: 'Became a paying customer',
      widthClass: 'w-20',
      color: '#ADD8E6',
    },
  ];

  beforeEach(() => {
    wrapper = mount(ChartDefaultFunnelBase, {
      props: {
        data: sampleData,
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
  });

  test('renders the correct number of funnel steps', () => {
    const funnelSteps = wrapper.findAll('.unnnic-chart-funnel-base-item');
    expect(funnelSteps).toHaveLength(sampleData.length);
  });

  test('renders the correct percentage and value for each step', () => {
    const funnelSteps = wrapper.findAll('.unnnic-chart-funnel-base-item');
    funnelSteps.forEach((step, index) => {
      expect(
        step.find('.unnnic-chart-funnel-base-item__text__values-title').text(),
      ).toBe(sampleData[index].percentage);
      expect(
        step
          .find('.unnnic-chart-funnel-base-item__text__values-sub-title')
          .text(),
      ).toBe(`| ${sampleData[index].value}`);
    });
  });

  test('renders the correct description for each step', () => {
    const funnelSteps = wrapper.findAll(
      '.unnnic-chart-funnel-base-item__text-description',
    );
    funnelSteps.forEach((step, index) => {
      expect(step.text()).toBe(sampleData[index].description);
    });
  });

  test('applies the correct width class for each step', () => {
    const funnelSteps = wrapper.findAll(
      '.unnnic-chart-funnel-base-item .overflow-hidden',
    );
    funnelSteps.forEach((step, index) => {
      expect(step.classes()).toContain(sampleData[index].widthClass);
    });
  });

  test('applies the first-item class to the first step', () => {
    const firstStep = wrapper.find(
      '.unnnic-chart-funnel-base-item .overflow-hidden.first-item',
    );
    expect(firstStep.exists()).toBe(true);
  });

  test('applies the last-item class to the last step', () => {
    const lastStep = wrapper.find(
      '.unnnic-chart-funnel-base-item .overflow-hidden.last-item',
    );
    expect(lastStep.exists()).toBe(true);
  });

  test('applies the last-item class to the text container of the last step', () => {
    const lastStepText = wrapper.find(
      '.unnnic-chart-funnel-base-item__text.last-item',
    );
    expect(lastStepText.exists()).toBe(true);
  });

  test('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
