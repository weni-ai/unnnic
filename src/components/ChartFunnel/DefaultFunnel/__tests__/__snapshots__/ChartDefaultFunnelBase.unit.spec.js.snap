// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChartDefaultFunnelBase.vue > matches the snapshot 1`] = `
"<section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-container">
  <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item">
    <section data-v-b3eefbe4="" class="overflow-hidden w-60 first-item">
      <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__card first-item" style="background-color: rgb(0, 119, 182);"></section>
    </section>
    <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text">
      <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values">
        <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values-title">90%</p>
        <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values-sub-title"> &nbsp; | 900</p>
      </section>
      <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text-description">Viewed the landing page</p>
    </section>
  </section>
  <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item">
    <section data-v-b3eefbe4="" class="overflow-hidden w-50">
      <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__card" style="background-color: rgb(0, 168, 232);"></section>
    </section>
    <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text">
      <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values">
        <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values-title">80%</p>
        <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values-sub-title"> &nbsp; | 800</p>
      </section>
      <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text-description">Clicked on the CTA</p>
    </section>
  </section>
  <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item">
    <section data-v-b3eefbe4="" class="overflow-hidden w-40">
      <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__card" style="background-color: rgb(0, 191, 255);"></section>
    </section>
    <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text">
      <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values">
        <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values-title">70%</p>
        <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values-sub-title"> &nbsp; | 700</p>
      </section>
      <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text-description">Signed up for the trial</p>
    </section>
  </section>
  <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item">
    <section data-v-b3eefbe4="" class="overflow-hidden w-30">
      <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__card" style="background-color: rgb(135, 206, 235);"></section>
    </section>
    <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text">
      <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values">
        <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values-title">60%</p>
        <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values-sub-title"> &nbsp; | 600</p>
      </section>
      <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text-description">Completed the onboarding</p>
    </section>
  </section>
  <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item">
    <section data-v-b3eefbe4="" class="overflow-hidden w-20 last-item">
      <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__card" style="background-color: rgb(173, 216, 230);"></section>
    </section>
    <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text last-item">
      <section data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values">
        <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values-title">50%</p>
        <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text__values-sub-title"> &nbsp; | 500</p>
      </section>
      <p data-v-b3eefbe4="" class="unnnic-chart-funnel-base-item__text-description">Became a paying customer</p>
    </section>
  </section>
</section>"
`;
